### ISSUE TITLE
Missing Key Support

### ISSUE THREAD

**Issue by Kethku** [Kethku/neovide#27]
Issue to track missing keybinding support

**Comment by unrealhoang** [comment]
```
[src/keybindings.rs:164] input = KeyboardInput {
    scancode: 29,
    state: Pressed,
    virtual_keycode: Some(
        LControl,
    ),
    modifiers: CTRL,
}
[src/keybindings.rs:164] input = KeyboardInput {
    scancode: 22,
    state: Pressed,
    virtual_keycode: Some(
        U,
    ),
    modifiers: ALT,
}
```
This is the keys recorded when I press `Ctrl-D`.

```
[src/keybindings.rs:164] input = KeyboardInput {
    scancode: 42,
    state: Pressed,
    virtual_keycode: Some(
        LShift,
    ),
    modifiers: SHIFT,
}
```
This is the keys recorded when I press Shift, nothing else recorded when I press other keys while Shift is pressed.

**Comment by Kethku** [comment]
I'm going to take another stab at your strategy for key input. It combined with the escaping of particular keys/handling of modifiers may be more consistent than what exists today.

**Comment by Kethku** [comment]
Postponing this to vnext. The problem is more complicated than previously anticipated. Just using the character received api ignores keys which do not produce a character. However as you have pointed out, just tracking the virtual key codes and translating does not take into account the individual users configuration. Probably the best solution is a mix of the two or waiting for https://github.com/rust-windowing/winit/issues/753 to be addressed. I think this is partially a fundamental limitation in winit...

**Comment by Kethku** [comment]
Note solving this will also require fixing modifier support for linux.

**Comment by osa1** [comment]
I wonder if this is a bug in skulpin (or one of its dependencies). If I print key events, when I press `C-l` in Linux I first get a Ctrl event as expected:

```
WindowEvent {
    window_id: WindowId(
        X(
            WindowId(
                37748737,
            ),
        ),
    ),
    event: KeyboardInput {
        device_id: DeviceId(
            X(
                DeviceId(
                    3,
                ),
            ),
        ),
        input: KeyboardInput {
            scancode: 97,
            state: Pressed,
            virtual_keycode: Some(
                RControl,
            ),
            modifiers: CTRL,
        },
        is_synthetic: false,
    },
}
```

However the event for `l` shows `ALT` modifier, not `CTRL`:

```
WindowEvent {
    window_id: WindowId(
        X(
            WindowId(
                37748737,
            ),
        ),
    ),
    event: KeyboardInput {
        device_id: DeviceId(
            X(
                DeviceId(
                    3,
                ),
            ),
        ),
        input: KeyboardInput {
            scancode: 38,
            state: Pressed,
            virtual_keycode: Some(
                L,
            ),
            modifiers: ALT,
        },
        is_synthetic: false,
    },
}
```

Not sure why that is, but it looks like a bug to me.

**Comment by Kethku** [comment]
The Neovim key input is very strange. Keybindings are sent to it as a string with the same format you use to map keybindings inside of vim. This has a couple of weird effects:

    there is a set of keys which don't provide text which need to be escaped with angle brackets (sometimes mapping the name to something different as well),
    some keys which do have clear text values also need to be mapped to a different name and escaped with angle brackets
    modifiers must be applied with a <M-{char}> or <C-{char}> format.
    for some keys specifically with shift key modification the expectation is that the modified key should be sent instead of <S-{key char}>. For example instead of sending the expectation is that we send ! directly.

So I thiiink the correct strategy is to use the ReceivedCharacter event with a small set of transforms for characters which must be escaped, use KeyboardInput for remaining keys which do not produce values, and use ModifiersChanged to track modifiers and append to the previous event values.

**Comment by Kethku** [comment]
This is supported by the fact that ModifiersChanged is the recommended method of reading modifier state since the modifier value I am currently using is now deprecated in Winit

**Comment by osa1** [comment]
@Kethku so I was looking at winit to see if this is a bug in winit, but it seems not. If I run [`window` example in winit][1] (which prints window events to stdout), I see this event when I press `Ctrl-l`:

```
WindowEvent { window_id: WindowId(X(WindowId(65011713))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 38, state: Released, virtual_keycode: Some(L), modifiers: CTRL }, is_synthetic: false } }
```

I wonder why this works when I use winit directly, but not in neovide (which I think uses winit via skulpin?).

[1]: https://github.com/rust-windowing/winit/blob/master/examples/window.rs

**Comment by osa1** [comment]
Indeed the modifier key bug is fixed in winit-0.20.0, see the changelog entry for it here: https://github.com/rust-windowing/winit/blob/master/CHANGELOG.md#0200-2020-01-05

Unfortunately skulpin uses a buggy version (0.20.0-alpha-6) which needs to be updated first.

**Comment by osa1** [comment]
It turns out skulpin can't really move to winit-0.20 because 0.20 introduces a bug on OSX. Locally when I use skulpin with winit-0.20 the issues with incorrect modifier keys are fixed.

**Comment by Kethku** [comment]
Ok interesting. Thats very encouraging that the swapped modifiers is an upstream change. I still need to overhaul the keyboard input management, but this explains my biggest confusion up till now. I may attempt it sometime tonight depending on how much time I get to work on it.

**Comment by Kethku** [comment]
Branch PR updating to a locked winit revision has been merged. Next step is to push a change which targets our skulpin to the winti_20 branch of skulpin. I don't have time at the moment but am willing to merge PRs. If not I will get to it later today.

**Comment by Kethku** [comment]
I just pushed a winit_20 branch which contains an updated skulpin and winit. There were some significant dpi changes, so I didn't want to just push it to master without people trying it

**Comment by Kethku** [comment]
Alright I manually tested it jumping between dpis, made some fixes and pushed to master. The upstream modifier bug should be fixed (not all of the keybinding problems are though)

**Comment by Kethku** [comment]
Unfortunately winit 0.20.0 crashes on both mac and windows, so this is going to have to wait a while longer.

**Comment by osa1** [comment]
Ugh... it's unfortunate that winit is so buggy. Btw I found a bug (or multiple bugs) in winit 0.20.0 too, some shift keys are working as expected, some do not. For example, Shift-L is working, but `%` (Shift + 5) is not...

**Comment by jonvaldes** [comment]
Regarding this issue, on Linux I'm seeing neovide completely ignore some keys on my keyboard (like the brackets keys)

Printing the events that get to neovide, I'm seeing this:
`KeyboardInput { scancode: 26, state: Released, virtual_keycode: None, modifiers: (empty) }
KeyboardInput { scancode: 27, state: Pressed, virtual_keycode: None, modifiers: (empty) }`

Neovide unfortunately relies entirely on the virtual_keycode, so any keypress that gets a null keycode will be ignored. This sounds like a winit bug to me?

**Comment by Kethku** [comment]
This is indeed a winit limitation.

Winit provides two input mechanisms:

The one we use which provides virtual keycodes and partial coverage of all key options

And recieve character which tries to parse the keyboard inputs and produce text. This covers more cases but does very weird things when modifiers are pressed. Further it supports some keys which don't produce text via control codes but not everything.

It has been on my list to go read up on how alacritty and servo do it but I haven't had the time or brain capacity to attempt it. Wezterm ended up rolling his own windowing system I believe to get around these issues which isn't a great sign. Further druid she'll decided to roll their own for the same reason.

There is some hope because winit has an open issue discussing a redesign but it has had mixed traction.

Things are pretty windows centric with neovide at the moment because I don't have access to a Linux box, but I may have to just buckle down and make one so I can test things more thoroughly.

Thanks for listening to my TED talk ;)

**Comment by jonvaldes** [comment]
So, both Servo and Alacritty interpret the ReceivedCharacter event (https://docs.rs/winit/0.20.0/winit/event/enum.WindowEvent.html#variant.ReceivedCharacter), and choose to supress it or not based on some internal logic: 
https://github.com/alacritty/alacritty/blob/74039bde1249ad77fe8d2d85ec6eb1f363f5c511/alacritty/src/input.rs#L607
https://github.com/servo/servo/blob/e07f0280fe9c16fbdeb34d488e00e8101f28fc9e/ports/glutin/headed_window.rs#L477

However, Servo then has a bunch of internal logic to interpret the scancodes and vkcodes:
https://github.com/servo/servo/blob/a9f7b132303f7c02648043df5ebdc25726bd3f88/ports/glutin/keyutils.rs
Alacritty doesn't seem to bother. Maybe we can ignore that too?

**Comment by jonvaldes** [comment]
Well, this hack does improve things: https://github.com/jonvaldes/neovide/commit/2c45503d4a99b658752490486b6a2649f854f3c8
It's not perfect, so things like `M-S-[` are actually interpreted as `M-S-{`, but... well, it does recognize the square brackets at least!

**Comment by Kethku** [comment]
That seems similar to what im thinking. One avenue I'm exploring is to look at how the control key modifies the received character and undo the modification. The resulting character can be sent directly to neovim and then any remaining gaps can be filled with the virtualkeycode

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by jonvaldes** [comment]
And a further improved hack: https://github.com/jonvaldes/neovide/commit/b4e1b039ee64a1e58fb507c7e50bcf5282803959
This works well enough for me to use, I think.
Thoughts?

**Comment by Kethku** [comment]
Interesting I will take a look shortly

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Kethku** [comment]
Wezterm does some crazy stuff and implements it for each platform https://github.com/wez/wezterm/tree/master/window/src/os

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Kethku** [comment]
@jonvaldes those changes seem like no brainers to me. I think they should be merged in as they demonstrably improve the situation. The long term fix is probably to wait on winit to merge something better. Until then I think small fixes like this that patch holes in the current input handling is a good strategy.

Could you make a PR with your branch so I can merge it in to master?

**Comment by jonvaldes** [comment]
Ok, turns out after my hack we're still missing 2 characters I can't type on Linux:
'+' and '<'.
Every other character I seem to be able to type fine. It's just those 2.

**Comment by Kethku** [comment]
I believe this should be fixed now with @jonvaldes wonderful work. The only remaining issue I know of is an upstream problem where remapped keys via autohotkey on windows or other mechanisms on linux don't get properly respected. So I'm going to close this and make a new issue to track that.

**Comment by Ninmi** [comment]
Along with + and <, I'm also not able to type ` and ^, possibly due to the fact that you need to press them twice to get the character? I'm on a Swedish/Finnish keyboard layout, if that matters.

One other weird quirk is how changing window with i3wm hotkey Super+j (for example) actually inserts <D-j> in the buffer. Only when you enter the window. But this probably goes on its own issue I guess?

**Comment by Kethku** [comment]
That is weird and should probably be its own issue. My knowledge of Linux and the window managers is severely lacking...

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Kethku** [comment]
The op of this issue should be fixed in the sdl2 branch.

**Comment by Kethku** [comment]
I believe this is fixed with the merging of sdl2

**Comment by ABitMoreDepth** [comment]
I still don't appear to be able to use Ctrl+6 (switch to previously open buffer), if thats something related to this issue?

**Comment by Kethku** [comment]
I'm not sure. Do you have a non-english keyboard layout? If so can you run with --log and upload the log after reproducing the issue? Also are you running from release or building yourself?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by ABitMoreDepth** [comment]
I am running a British keyboard layout currently (though I use a US layout at work)
I notice that the number 6 appears to the bottom right of the UI:
![image](https://user-images.githubusercontent.com/3308997/76146406-608f9f80-608a-11ea-8beb-6bcfdb91e1f4.png)

Looking at the logs, the relevant bits are here:
```
TRACE [neovide::window] KeyDown Received: Left Ctrl
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
TRACE [neovide::window] KeyDown Received: 6
TRACE [neovide::bridge] UiCommand queued: Keyboard("<C-6>")
DEBUG [neovide::window] Render Triggered
TRACE [neovide::bridge] Executing UiCommand: Keyboard("<C-6>")
TRACE [neovide::bridge::ui_commands] Keyboard Input Sent: <C-6>
TRACE [neovide::bridge::handler] Neovim notification: "redraw"
TRACE [neovide::editor] Image flushed
TRACE [neovide::redraw_scheduler] Next frame queued
```

So it certainly looks like its receiving the key information, but in neovim, and neovim-gtk, the same keybinding swaps between open buffers?  I note that in `:help alternate-file` it refers to the appropriate keybind as `<C-^>`, so I tried using Ctrl+Shit+6, but that behaves in the same manner.

**Comment by Kethku** [comment]
Can you show me the log from ctrl-shift-6?

**Comment by ABitMoreDepth** [comment]
```
TRACE [neovide::window] KeyDown Received: Left Ctrl
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
TRACE [neovide::window] KeyDown Received: Left Shift
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
TRACE [neovide::window] KeyDown Received: 6
TRACE [neovide::bridge] UiCommand queued: Keyboard("<C-S-6>")
DEBUG [neovide::window] Render Triggered
TRACE [neovide::bridge] Executing UiCommand: Keyboard("<C-S-6>")
TRACE [neovide::bridge::ui_commands] Keyboard Input Sent: <C-S-6>
TRACE [neovide::bridge::handler] Neovim notification: "redraw"
TRACE [neovide::editor] Image flushed
TRACE [neovide::redraw_scheduler] Next frame queued
DEBUG [neovide::window] Render Triggered
TRACE [neovide::renderer] Rendering
TRACE [neovide::editor] Draw commands sent
DEBUG [neovide::window] Render Triggered
```

**Comment by Kethku** [comment]
Ah yup this is a live one. I'm brainstorming strategies for how to fix this...

**Comment by svermeulen** [comment]
I noticed after upgrading today that all command key mappings (ie. the ones starting with `<d-`) do not work anymore.  This is on macOS catalina 10.15.3

**Comment by Kethku** [comment]
Argh my bad :( I'll fix this

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by ABitMoreDepth** [comment]
Happy to provide testing etc for you if desired, thanks for looking into things!

**Comment by Kethku** [comment]
@svermeulen I fixed the d bindings. It doesn't work on windows since the windows key is used for other things, but it should work on linux and mac.

**Comment by svermeulen** [comment]
Yep that fixed it  :)  Thanks

**Comment by Kethku** [comment]
Aight. At least for qwerty I believe the above is fixed. I did it by writing out all of the possible keyboard modifier combinations for a standard qwerty keyboard. This is necessary because the sdl2 textinput event doesn't produce values when ctrl is held down on windows machines. I've attempted to contact the sdl folks with no luck.

So for now I created a setting variable `g:neovide_keyboard_layout` which has one valid value at the moment: "qwerty". Eventually and depending on demand. I will add other keyboard layouts in the same way.

Ideally we would rely on OS support for this, but until either Winit becomes less buggy or an SDL solution becomes known, the manual approach will be necessary.

@ABitMoreDepth could you test this out and let me know if it works for you? Your suggested keybinding of <C-6> never worked for me, but <C-^> (the modified version of <C-S-6>) should work on master now.

**Comment by ABitMoreDepth** [comment]
Looks like your adjustments work for `<C-^>`!, However I am very accustomed to using (the keyboard equivalent of `<C-6>`, which currently doesn't switch buffers:

```
TRACE [neovide::window] Keyboard Input Received: keycode-Some(LCtrl) modifiers-LCTRLMOD text-None
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
TRACE [neovide::window] Keyboard Input Received: keycode-Some(Num6) modifiers-LCTRLMOD text-None
TRACE [neovide::bridge] UiCommand queued: Keyboard("<C-6>")
DEBUG [neovide::window] Render Triggered
TRACE [neovide::bridge] Executing UiCommand: Keyboard("<C-6>")
TRACE [neovide::bridge::ui_commands] Keyboard Input Sent: <C-6>
TRACE [neovide::bridge::handler] Neovim notification: "redraw"
TRACE [neovide::editor] Image flushed
TRACE [neovide::redraw_scheduler] Next frame queued
DEBUG [neovide::window] Render Triggered
TRACE [neovide::renderer] Rendering
TRACE [neovide::editor] Draw commands sent
```

Pressing `<C-^>` (Ctrl-Shift-Six, on my keyboard):
```
TRACE [neovide::window] Keyboard Input Received: keycode-Some(LCtrl) modifiers-LCTRLMOD text-None
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
TRACE [neovide::window] Keyboard Input Received: keycode-Some(LShift) modifiers-LSHIFTMOD | LCTRLMOD text-None
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
TRACE [neovide::window] Keyboard Input Received: keycode-Some(Num6) modifiers-LSHIFTMOD | LCTRLMOD text-None
TRACE [neovide::bridge] UiCommand queued: Keyboard("<C-^>")
DEBUG [neovide::window] Render Triggered
TRACE [neovide::bridge] Executing UiCommand: Keyboard("<C-^>")
TRACE [neovide::bridge::ui_commands] Keyboard Input Sent: <C-^>
TRACE [neovide::bridge::handler] Neovim notification: "redraw"
```

Thinking more about it, your adjustment seems correct, digging around in the vim help, I found this:
```
                                                        CTRL-^ CTRL-6
CTRL-^                  Edit the alternate file.  Mostly the alternate file is
                        the previously edited file.  This is a quick way to
                        toggle between two files.  It is equivalent to ":e #",
                        except that it also works when there is no file name.

                        If the 'autowrite' or 'autowriteall' option is on and
                        the buffer was changed, write it.
                        Mostly the ^ character is positioned on the 6 key,
                        pressing CTRL and 6 then gets you what we call CTRL-^.
                        But on some non-US keyboards CTRL-^ is produced in
                        another way.
```

The help seems somewhat ambiguous as to the correct behavior, and some suggest that Ctrl-6 works more by accident (though it is explicitly included in tools like GVim).

As things stand, I can add the following line to my vim config to match previously observed behavior:
```vim
nnoremap <C-6> <C-^>
```

I'm happy enough with the above fix, it seems slightly uncertain what the originally desired behavior was.

Thanks for finding a fix!  Its unfortunate you've had to manually create a mapping though, hopefully the various libraries can implement proper support for modifiers soon.!

**Comment by svermeulen** [comment]
Another key issue I just noticed:  When hitting the `>` key in normal mode it is triggered as `<S->>` instead of just `>`

**Comment by Kethku** [comment]
Ah yup. Good catch. I've fixed this particular issue. However there are still other problems

I'm running out of ideas. On the one hand, I don't think manually writing out keyboard layouts is a sustainable solution as we will always have an incompatibility. On the other hand, the sdl2 api is just not sufficient. The sdl2 folks closed my bug https://bugzilla.libsdl.org/show_bug.cgi?id=5029 saying that the textinput api is just for text and should not be used for keybindings. This makes sense, but in the vim context, it doesn't work because keybindings are set using the actual produced text, not the key name.

So the current state of the world is:

1.  I will add keyboard layouts as they are requested to the list which should allow them to set the keyboard layout in their vim config (I don't like this, but I don't see a way around it)
2.  If a keyboard layout requires ctrl or alt to input a base key, it will be reported as <C-{that key}> or <M-{that key}>. This should work in most cases, but might cause issues in the future

This gets as as close as I think we're going to get with the current crossplatform keyboard input support. To make testing out other windowing libraries easier (and to make abstracting the renderer easier) I am going to soon create a trait which represents the windowing library. That way we can experiment with sdl2 and winit support without swapping the whole world out. Hopefully sometime soon we can swap back to winit which might make the keyboard situation better... We'll see.

For now though I'm going to close this. If there are more issues though please let me know and reopen. I'll keep iterating or tell you if its not possible with the current libraries
